lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -771,6 +766,121 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
+    -> IO ( Observer IO W.ChimericAccount W.Coin
+          , Tip (CardanoBlock sc) -> IO ()
+            -- Call on tip-change to refresh
+          )
+newRewardBalanceFetcher tr gp queryRewardQ = newObserver nullTracer fetch
+  where
+    fetch
+        :: Tip (CardanoBlock sc)
+        -> Set W.ChimericAccount
+        -> IO (Maybe (Map W.ChimericAccount W.Coin))
+    fetch tip accounts = do
+        liftIO $ traceWith tr $
+            MsgGetRewardAccountBalance (fromTip' gp tip) accounts
+        let creds = Set.map toStakeCredential accounts
+        let q = QueryIfCurrentShelley (Shelley.GetFilteredDelegationsAndRewardAccounts creds)
+        let cmd = CmdQueryLocalState (getTipPoint tip) q
+
+        res <- liftIO . timeQryAndLog "getAccountBalance" tr $
+            queryRewardQ `send` cmd
+        case res of
+            Right (Right (deleg, rewardAccounts)) -> do
+                liftIO $ traceWith tr $ MsgAccountDelegationAndRewards deleg rewardAccounts
+                return $ Just $ Map.mapKeys fromStakeCredential
+                    $ Map.map fromShelleyCoin rewardAccounts
+
+            Right (Left _) -> -- wrong era
+                return
+                    . Just
+                    . Map.fromList
+                    . map (, minBound)
+                    $ Set.toList accounts
+            Left acqFail -> do
+                -- NOTE: this could possibly happen in rare circumstances when
+                -- the chain is switched and the local state query is made
+                -- before the node tip variable is updated.
+                liftIO $ traceWith tr $
+                    MsgLocalStateQueryError DelegationRewardsClient $
+                    show acqFail
+                return Nothing
+
+
+data ObserverLog key value
+    = MsgWillFetch (Set key)
+    | MsgDidFetch (Map key value)
+    | MsgAddedObserver key
+    | MsgRemovedObserver key
+
+-- | Given a way to fetch values for a set of keys, create:
+-- 1. An @Observer@ for consuming values
+-- 2. A refresh action
+--
+-- The @env@ parameter can be used to pass in information needed for refreshing,
+-- like the current tip when fetching rewards.
+newObserver
+    :: forall m key value env. (MonadSTM m, Ord key)
+    => Tracer m (ObserverLog key value)
+    -> (env -> Set key -> m (Maybe (Map key value)))
+    -> m (Observer m key value, env -> m ())
+newObserver tr fetch = do
+    cacheVar <- atomically $ newTVar Map.empty
+    toBeObservedVar <- atomically $ newTVar Set.empty
+    return (observer cacheVar toBeObservedVar, refresh cacheVar toBeObservedVar)
+  where
+    observer
+        :: TVar m (Map key value)
+        -> TVar m (Set key)
+        -> Observer m key value
+    observer cacheVar observedKeysVar =
+        Observer
+            { startObserving = \k -> do
+                atomically $ do
+                    modifyTVar' observedKeysVar (Set.insert k)
+                traceWith tr $ MsgAddedObserver k
```

Comment:
Since "startObserving" is called every time `getAccountBalance` is called, this would log a line `MsgAddedObserver` in a rather misleading way and much more than it needs. We could use something like `Set.alterF` and return `Just _` or `Nothing` depending on whether the element was added (and log accordingly).

Reply:
Yup. It's connected to `nullTracer` now, so one doesn't notice, but should connect to the real tracer and change this.